File: src/xingcheng/rules.py
"""Explicit source-scoped rules and semester expansion; never infer holidays."""

from copy import deepcopy
from datetime import date, timedelta
import re


def clock(value):
    if not isinstance(value, str) or not re.fullmatch(
        r"(?:[01]\d|2[0-3]):[0-5]\d", value
    ):
        raise ValueError("时间应为 HH:MM")
    return value


def validate_rules(rules):
    if not isinstance(rules, dict) or set(rules) - {"shifts", "semester", "template"}:
        raise ValueError("规则配置无效")
    shifts = rules.get("shifts", [])
    if not isinstance(shifts, list) or len(shifts) > 100:
        raise ValueError("最多配置 100 个班次")
    aliases = set()
    for shift in shifts:
        if not isinstance(shift, dict) or set(shift) - {
            "name",
            "aliases",
            "start",
            "end",
            "next_day",
        }:
            raise ValueError("班次模板包含不支持的字段")
        if (
            not isinstance(shift.get("name"), str)
            or len(shift["name"]) > 100
            or not isinstance(shift.get("aliases"), list)
            or len(shift["aliases"]) > 100
        ):
            raise ValueError("请填写有效的班次名称和符号列表")
        if "next_day" in shift and not isinstance(shift["next_day"], bool):
            raise ValueError("次日结束应为明确的开关设置")
        if not shift.get("name") or not shift.get("aliases"):
            raise ValueError("请填写班次名称和符号")
        for alias in shift["aliases"]:
            if (
                not isinstance(alias, str)
                or not alias.strip()
                or len(alias) > 100
                or alias in aliases
            ):
                raise ValueError("班次符号重复或为空")
            aliases.add(alias)
        clock(shift["start"])
        clock(shift["end"])
        if not shift.get("next_day") and shift["end"] <= shift["start"]:
            raise ValueError("跨夜班次请设置次日结束")
    semester = rules.get("semester")
    if semester:
        if (
            not isinstance(semester, dict)
            or set(semester) - {"monday", "weeks", "periods"}
            or not isinstance(semester.get("periods"), list)
        ):
            raise ValueError("学期模板格式无效")
        if date.fromisoformat(semester["monday"]).weekday() != 0:
            raise ValueError("第一教学周必须选择周一")
        if not isinstance(semester["weeks"], int) or not 1 <= semester["weeks"] <= 60:
            raise ValueError("学期周数应在 1–60 之间")
        for slot in semester["periods"]:
            if not isinstance(slot, dict) or set(slot) != {"start", "end"}:
                raise ValueError("节次仅包含开始和结束时间")
            clock(slot["start"])
            clock(slot["end"])
            if slot["end"] <= slot["start"]:
                raise ValueError("节次结束时间必须晚于开始时间")
        if not 1 <= len(semester["periods"]) <= 30:
            raise ValueError("请设置 1–30 个节次")
        if any(
            b["start"] < a["end"]
            for a, b in zip(semester["periods"], semester["periods"][1:])
        ):
            raise ValueError("节次必须按时间排序且不重叠")
    template = rules.get("template")
    if template:
        if not isinstance(template, dict) or set(template) - {
            "layout",
            "sheet",
            "header_row",
            "headers",
            "mapping",
        }:
            raise ValueError("表格模板包含不支持的字段")
        if template.get("layout") not in ("records", "names_rows", "names_columns"):
            raise ValueError("模板布局无效")
        if (
            not isinstance(template.get("header_row"), int)
            or template["header_row"] < 0
        ):
            raise ValueError("请指定表头行")
        if not template.get("headers") or not isinstance(template.get("mapping"), dict):
            raise ValueError("请保存表头特征和字段映射")
        if not isinstance(template["headers"], list) or any(
            not isinstance(h, str) for h in template["headers"]
        ):
            raise ValueError("表头必须是文字列表")
        fields = {
            "name",
            "date",
            "title",
            "shift",
            "start",
            "end",
            "time",
            "location",
            "note",
        }
        if set(template["mapping"]) - fields or any(
            not isinstance(v, int) or v < 0 for v in template["mapping"].values()
        ):
            raise ValueError("字段映射必须使用有效的行列位置")
    return deepcopy(rules)


def apply_rules(value, rules):
    validate_rules(rules)
    result = deepcopy(value)
    # Original times always take precedence, including an explicit start-only time.
    if result.get("start") or not result.get("date"):
        return result
    alias = result.get("shift") or result.get("title")
    rule = next((s for s in rules.get("shifts", []) if alias in s["aliases"]), None)
    if rule:
        result.update(
            start=rule["start"],
            end=rule["end"],
            precision="interval",
            end_date=(
                date.fromisoformat(result["date"])
                + timedelta(days=int(bool(rule.get("next_day"))))
            ).isoformat(),
        )
        result.setdefault("field_basis", {}).update(
            {k: "个人规则：" + rule["name"] for k in ("start", "end", "end_date")}
        )
    return result


def expand_course(course, semester):
    validate_rules({"semester": semester})
    weekday = course["weekday"]
    periods = course["periods"]
    if (
        not isinstance(weekday, int)
        or not 1 <= weekday <= 7
        or not periods
        or periods != list(range(min(periods), max(periods) + 1))
    ):
        raise ValueError("请选择星期和连续节次")
    if min(periods) < 1 or max(periods) > len(semester["periods"]):
        raise ValueError("节次超出配置范围")
    weeks = course.get("weeks") or list(
        range(course.get("week_from", 1), course.get("week_to", semester["weeks"]) + 1)
    )
    if (
        not weeks
        or not isinstance(course.get("title"), str)
        or not course["title"].strip()
    ):
        raise ValueError("请填写课程名称和有效周次")
    if any(not isinstance(w, int) or not 1 <= w <= semester["weeks"] for w in weeks):
        raise ValueError("周次超出学期范围")
    parity = course.get("parity", "all")
    if parity not in ("all", "odd", "even"):
        raise ValueError("单双周设置无效")
    from .calendar import normalized

    events = []
    for week in sorted(set(weeks)):
        if parity == "odd" and week % 2 == 0 or parity == "even" and week % 2 == 1:
            continue
        day = (
            date.fromisoformat(semester["monday"])
            + timedelta(weeks=week - 1, days=weekday - 1)
        ).isoformat()
        events.append(
            normalized(
                dict(
                    date=day,
                    title=course["title"],
                    location=course.get("location", ""),
                    category="学习",
                    start=semester["periods"][min(periods) - 1]["start"],
                    end=semester["periods"][max(periods) - 1]["end"],
                    end_date=day,
                    sources=course.get("sources", []),
                    notes=[f"第 {week} 教学周"],
                    field_basis={k: "个人课程规则" for k in ("date", "start", "end")},
                )
            )
        )
    return dict(events=events, pending=[], files=[], warnings=[], conflicts=[])

File: miniprogram/pages/rules/rules.js
const api = require('../../utils/api');
const store = require('../../utils/calendar-store');
Page({
  data: { sources: [], index: 0, error: '', busy: false, name: '', aliases: '', shiftName: '', start: '', end: '', nextDay: false, shifts: [],
    monday: '', weeks: 20, periodsText: '', courseTitle: '', location: '', weekday: 1, periodFrom: 1, periodTo: 2, weekFrom: 1, weekTo: 20,
    parity: 0, parityLabels: ['全部周', '单周', '双周'], specificWeeks: '', templateLayout: 0, layouts: ['逐行明细表', '姓名在行、日期在列', '姓名在列、日期在行'], headerRow: 1,
    nameColumn: 1, dateColumn: 2, titleColumn: 3, timeColumn: 4, templateSheet: '', preview: null },
  onLoad(options) {
    const c = getApp().globalData.calendar;
    if (!c) return this.setData({ error: '请先建立个人日历' });
    this._calendar = c;
    const sources = c.sources.map(s => ({ id: s.id, name: s.name }));
    this.setData({ sources, index: Math.max(0, sources.findIndex(s => s.id === options.id)) }); this.loadSource();
  },
  loadSource() {
    const s = this._calendar.sources[this.data.index];
    if (!s) return;
    const rules = s.rules || {}, semester = rules.semester || {};
    this._rules = JSON.parse(JSON.stringify(rules));
    this.setData({ name: s.name, shifts: rules.shifts || [], monday: semester.monday || '', weeks: semester.weeks || 20,
      periodsText: (semester.periods || []).map(p => `${p.start}-${p.end}`).join('\n'), templateSaved: !!rules.template, preview: null,
      revisions:s.revisions.map(r=>({id:r.id,importedAt:r.imported_at,coverage:r.coverage ? r.coverage.join(' 至 ') : '追加导入',files:r.report.files.map(f=>f.filename).join('、')})) });
    this._sheets = s.revisions.flatMap(r => r.report.files.flatMap(f => f.sheets)).filter(s => s.preview && s.preview.length);
    this.setData({ tableSheets: this._sheets.map((s, i) => ({ label: `${i + 1}. ${s.sheet}` })), tablePreview: this._sheets[0] ? this._sheets[0].preview : [], sheetIndex: 0 });
  },
  source(event) { this.setData({ index: Number(event.detail.value) }); this.loadSource(); },
  field(event) { this.setData({ [event.currentTarget.dataset.field]: event.detail.value, preview: null }); if(event.currentTarget.dataset.field==='sheetIndex') this.setData({tablePreview:this._sheets[Number(event.detail.value)].preview}); this._preview = null; },
  addShift() {
    const s = { name: this.data.shiftName.trim(), aliases: this.data.aliases.split(/[、,，/]/).map(s => s.trim()).filter(Boolean), start: this.data.start, end: this.data.end, next_day: this.data.nextDay };
    if (!s.name || !s.aliases.length || !s.start || !s.end) return this.setData({ error: '请填写班次名称、符号和起止时间' });
    this.setData({ shifts: [...this.data.shifts, s], preview: null, error: '' }); this._preview = null;
  },
  removeShift(event) { this.setData({ shifts: this.data.shifts.filter((_, i) => i !== Number(event.currentTarget.dataset.index)), preview: null }); this._preview = null; },
  config() {
    const rules = { ...(this._rules || {}), shifts: this.data.shifts };
    if (this.data.monday) {
      const periods = this.data.periodsText.trim().split('\n').filter(Boolean).map(line => { const p = line.trim().split(/[-–]/); return { start: (p[0] || '').trim(), end: (p[1] || '').trim() }; });
      rules.semester = { monday: this.data.monday, weeks: Number(this.data.weeks), periods };
    }
    return rules;
  },
  template() {
    const sheet = (this._sheets || [])[Number(this.data.sheetIndex)];
    const header = sheet && sheet.preview[Number(this.data.headerRow) - 1];
    if (!header) return this.setData({ error: '请选择包含表头的来源版本和表头行（局部预览前 20 行）' });
    this._rules = { ...(this._rules || {}), template: { sheet: sheet.sheet, header_row: Number(this.data.headerRow) - 1,
      headers: header.map(c => c.text), layout: ['records', 'names_rows', 'names_columns'][this.data.templateLayout],
      mapping: { name: Number(this.data.nameColumn) - 1, date: Number(this.data.dateColumn) - 1, title: Number(this.data.titleColumn) - 1, time: Number(this.data.timeColumn) - 1 } } };
    this.setData({ templateSaved: true, preview: null, error: '模板已加入草稿，请预览后保存' }); this._preview = null;
  },
  async preview() {
    if (this.data.busy || !this._calendar) return;
    const source = this._calendar.sources[this.data.index];
    if (!source) return this.setData({ error: '请先导入一个来源，再为它配置规则' });
    this.setData({ busy: true, error: '' });
    const fingerprint = JSON.stringify([this.data.index,this.data.name,this.config()]);
    try { const result = await api.transform(this._calendar, { type: 'source_config', source_id: source.id, name: this.data.name, rules: this.config(), apply: true });
      if (fingerprint !== JSON.stringify([this.data.index,this.data.name,this.config()])) throw Error('规则已变化，请重新预览');
      this._preview = result; this.setData({ preview: result.summary.changed.map(c => ({ id: c.id, before: `${c.before.date} ${c.before.start || '时间未注明'} ${c.before.title}`, after: `${c.after.date} ${c.after.start || '时间未注明'}–${c.after.end || ''} ${c.after.title}` })) });
    } catch (error) { this.setData({ error: error.message }); } finally { this.setData({ busy: false }); }
  },
  save() {
    if (!this._preview) return;
    try { store.commit(getApp().globalData, this._preview, getApp().globalData.referenceYear); this._calendar = getApp().globalData.calendar; this.loadSource(); this.setData({ error: '规则已保存' }); }
    catch (error) { this.setData({ error: error.message }); }
  },
  async course() {
    if (!this._calendar || this.data.busy) return;
    this.setData({ busy: true, error: '' });
    try {
      const rules = this.config();
      if (!rules.semester) throw Error('请设置第一教学周周一和节次时间');
      const a = Number(this.data.periodFrom), b = Number(this.data.periodTo);
      if (!Number.isInteger(a) || !Number.isInteger(b) || b < a || b - a > 30) throw Error('连续节次范围无效');
      const course = { title: this.data.courseTitle, location: this.data.location, weekday: Number(this.data.weekday), periods: Array.from({length: b - a + 1}, (_, i) => a + i),
        week_from: Number(this.data.weekFrom), week_to: Number(this.data.weekTo), parity: ['all', 'odd', 'even'][this.data.parity],
        weeks: this.data.specificWeeks ? this.data.specificWeeks.split(/[,，、\s]+/).filter(Boolean).map(Number) : undefined };
      const report = await api.request('/api/calendar/course', { course, semester: rules.semester });
      getApp().globalData.draft = { report, rules, name: this._calendar.personName, year: new Date().getFullYear(), sourceName: `${this.data.courseTitle}课程` };
      wx.navigateTo({ url: '/pages/confirm/confirm' });
    } catch (error) { this.setData({ error: error.message }); } finally { this.setData({ busy: false }); }
  }
});
